### DISCUSSION TITLE
FrozenResult serialization/deserialization

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#8006] [Usage Questions]
Hello, so I'm looking into the `FrozenResult` implementation. I know it's meant for caching. Are there specific steps we have to take to make it serializable? `pickle` should work with it just fine. However, to use it with other serialization techniques, what steps do we need to follow?

**Comment by zzzeek** [comment]
it's too bad because I'd love to know where the pickle bug is, that's very likely on our end.

FrozenResult is basically a list of tuples, or for some ORM cases a list of objects.  The list of tuples / objects is inside the ".data" collection.     Whether or not .data is a list of tuples or a list of objects is based on the FrozenResult._source_supports_scalars attribute.

There's a method rewrite_rows() that will give you all the data as a list of lists, if you don't want to look at _source_supports_scalars.     

basically you would serialize frozen_result.metadata, which is likely still best done using pickle, and then serialize frozen_result.rewrite_rows(), the list of lists.

There's not a public method to re-generate the FrozenResult.  I mean, it's a really simple object and if you look at the [source](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/engine/result.py#L1999) there's not much to it.  but if you are trying to avoid underscore methods, you would likely use that "metadata" and "rewritten rows" and pass it to a new IteratorResult.

```py
from sqlalchemy.engine.result import IteratorResult

metadata = my_frozen_result.metadata
plain_data = my_frozen_result.rewrite_rows()

new_result = IteratorResult(metadata, iter(plain_data))
```

however this is interesting.   maybe do the above and still use pickle for each row / column at a time.   Because then you will have a routine that nails down what object the pickle is failing on.

**Reply by michaelbukachi** [reply]
I'm still investigating the pickling error. I'm working on a minimum reproducible example with the same amount of models as is in our codebase. I'll post in the discussion as soon as I'm done.
